packages/studio-be/src/core/migration/bot-migration-service.ts

```diff
@@ -68,58 +58,66 @@ export class BotMigrationService {
-{bold ${centerText(`Migration${migrations.length === 1 ? '' : 's'} Required`, 40, 9)}}
-{dim ${centerText(`Version ${configVersion} => ${this.migService.targetVersion} `, 40, 9)}}
-{dim ${centerText(`${migrations.length} change${migrations.length === 1 ? '' : 's'}`, 40, 9)}}
-${_.repeat(' ', 9)}========================================`)
+    return missing.filter(x => x !== undefined) as MigrationFile[]
+  }
 
-    Object.keys(types).map(type => {
-      logger.warn(chalk`{bold ${types[type]}}`)
-      const filtered = migrations.filter(x => x.type === type)
+  private writeLabel(migrations: number, botId: string, version: string) {
+    const configLabel = `(${version} => ${this.migService.targetVersion})`
+    const changesLabel = `${migrations} change${migrations === 1 ? '' : 's'}`
 
-      if (filtered.length) {
-        filtered.map(x => logger.warn(`- ${x.description}`))
-      } else {
-        logger.warn('- None')
-      }
-    })
+    return chalk`{bold Migration${migrations === 1 ? '' : 's'} Required for ${botId}} ${configLabel} - ${changesLabel}`
   }
 
-  private async executeBotMigrations(botId: string, missingMigrations: MigrationFile[]) {
-    this.logger.info(chalk`
-${_.repeat(' ', 9)}========================================
-{bold ${centerText(
-      `Executing ${missingMigrations.length} migration${missingMigrations.length === 1 ? '' : 's'}`,
-      40,
-      9
-    )}}
-${_.repeat(' ', 9)}========================================`)
-
-    const opts = await this.migService.getMigrationOpts({ botId })
+  private async executeBotMigrations(botId: string, migrations: MigrationFile[], configVersion: string) {
+    const botConfig = await this.configProvider.getBotConfig(botId)
+    const opts = await this.migService.getMigrationOpts({ botId, botConfig })
     let hasFailures = false
 
-    await Promise.mapSeries(missingMigrations, async ({ filename }) => {
-      const result = await this.migService.loadedMigrations[filename].up(opts)
-      debug.forBot(botId, 'Migration step finished', { filename, result })
-      if (result.success) {
-        this.logger.forBot(botId).info(`- ${result.message || 'Success'}`)
-      } else {
-        hasFailures = true
-        this.logger.forBot(botId).error(`- ${result.message || 'Failure'}`)
+    const logger = this.logger.forBot(botId)
+
+    logger.warn(this.writeLabel(migrations.length, botId, configVersion))
+
+    await Promise.mapSeries(migrations, async ({ filename, info }) => {
+      const label = `${types[info.type]} - ${info.description}`
+
+      try {
+        const result = await this.migService.loadedMigrations[filename].up(opts)
+        debug.forBot(botId, 'Migration step finished', { filename, result })
+
+        if (result.success) {
+          logger.warn(`${chalk.green('[success]')} ${label}`)
+        } else if (result.hasChanges) {
+          logger.error(`${chalk.red('[failure]')} ${label}: ${result.message || ''}`)
```

Comment:
`hasChanges` means `failure`?

Reply:
If the request was not successful AND there were changes that had to be done, then yes it's a failure

Reply:
Alright, and if there were no changes but some errors, we are okay with it? I think it would be a lot simpler if we only had two kind of results: success or failure.

Reply:
if there were no changes then there were no errors, actually i'm not even sure we have migrations returning "success: false" (they would fall into the exception handler)